## 1. Load data from web page

Scrape the Wikipedia page https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

In [2]:
tables = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
postal_codes = tables[0]
postal_codes.head()

Postal code           Borough                Neighborhood
0         M1A      Not assigned                         NaN
1         M2A      Not assigned                         NaN
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

Filter dataframe deleting rows containing Not assigned Postal Codes. Rename first column.

In [3]:
postal_codes = postal_codes[postal_codes['Borough'] != 'Not assigned'].reset_index(drop=True)
postal_codes.rename(columns={'Postal code': 'PostalCode'}, inplace=True)

postal_codes.head()

PostalCode           Borough                                  Neighborhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In order to check duplicate postal codes we can count unique items for each column.

In [4]:
postal_codes.describe()

PostalCode     Borough Neighborhood
count         103         103          103
unique        103          10           98
top           M5M  North York    Downsview
freq            1          24            4

No duplicate postal codes! It seems thet neighborhoods are already grouped by postal code.<br/>
But some neighborhoods are listed twice, since we have 98 unique neighborhoods over 103 postal codes. We will handle this issue later.<br/>
Let's check if there is any NaN or Not Assigned in Neighborhood column.

In [5]:
postal_codes['Neighborhood'].isna().sum()

0

In [6]:
(postal_codes['Neighborhood']=='Not assigned').sum()

0

Multiple Neighborhoods for the same Postal Code are formatted with '/', replace it with a single comma (,).

In [7]:
postal_codes['Neighborhood'].replace(' /',',',regex=True,inplace=True)
postal_codes.head()

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [8]:
postal_codes.shape

(103, 3)

The first part of the project is complete.

## 2. Build Dataset

Let's add coordinates columns to the dataframe.

In [9]:
postal_codes['Latitude'] = 0.0
postal_codes['Longitude'] = 0.0

postal_codes.head()

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

   Latitude  Longitude  
0       0.0        0.0  
1       0.0        0.0  
2       0.0        0.0  
3       0.0        0.0  
4       0.0        0.0

Let's find out which neighborhoods appear more than once.

In [10]:
postal_codes_grouped = postal_codes.groupby('Neighborhood').count()

postal_codes_grouped[postal_codes_grouped['PostalCode']>1]

PostalCode  Borough  Latitude  Longitude
Neighborhood                                          
Don Mills              2        2         2          2
Downsview              4        4         4          4
Willowdale             2        2         2          2

In next dataframe iteration we will append Postal Code to these Neighborhoods in order to have unique Neighborhoods

Use geocoder to get coordinates and store them into dataframe. Use arcgis provider instead of google.

In [12]:
! conda install -c conda-forge geocoder

import geocoder # import geocoder

print('geocoder imported!')

geocoder imported!


In [13]:
# iterate over dataframe and add values for latitude and longitude
for index, row in postal_codes.iterrows():
    postal_code = row['PostalCode']
    g = geocoder.arcgis('{}, Toronto, Ontario'.format(postal_code))
    postal_codes.at[index,'Latitude']= g.lat
    postal_codes.at[index,'Longitude'] = g.lng
    if  row['Neighborhood'] in ['Don Mills','Downsview','Willowdale']:
        postal_codes.at[index,'Neighborhood'] = postal_codes.at[index,'Neighborhood'] + ' ' + postal_codes.at[index,'PostalCode']
    
# postal_codes contains all neighborhoods information now

Let's take a look at the dataframe now

In [14]:
postal_codes.shape

(103, 5)

In [15]:
postal_codes.head()

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.752935 -79.335641  
1  43.728102 -79.311890  
2  43.650964 -79.353041  
3  43.723265 -79.451211  
4  43.661790 -79.389390

Dataframe is now ready for further analysis!


## 3. Analyze Data

### 3.1 Explore Neighborhoods

Import libraries first.

In [16]:
import json # library to handle JSON files

import requests # library to handle requests

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


Get Toronto coordinates and create a map with neighborhoods

In [17]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geographical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geographical coordinates of Toronto are 43.6534817, -79.3839347.


In [18]:
# create map of Toronto using latitude and longitude values
map_toronto_full = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(postal_codes['Latitude'], postal_codes['Longitude'], postal_codes['Borough'], postal_codes['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto_full)  
    
map_toronto_full

In [19]:
toronto_data = postal_codes
toronto_data.shape

(103, 5)

Foursquare Credentials

In [20]:
CLIENT_ID = 'QJ1PBI3IOFN5VJL5UCNGYC5NM5JVEXMUJP5VLXIE0V4VAT4S' #  Foursquare ID
CLIENT_SECRET = 'EVQVSSLLIO0RMLFEZMGBRZ2PTE4V2O2ZZ2JYDDYMOPZ5HCAP' #  Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('credentials:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

credentials:
CLIENT_ID: QJ1PBI3IOFN5VJL5UCNGYC5NM5JVEXMUJP5VLXIE0V4VAT4S
CLIENT_SECRET:EVQVSSLLIO0RMLFEZMGBRZ2PTE4V2O2ZZ2JYDDYMOPZ5HCAP


Define Functions

In [21]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']
  
#
# function that gets top 100 venues for each neighborhood
# names, latitudes, longitudes are the columns of the dataframe
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now let's build a new dataframe with all the venues

In [22]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude'])

Let's check the results

In [23]:
print(str(toronto_venues.shape[0]) + ' venues found in ' + str(toronto_venues.groupby('Neighborhood')['Neighborhood'].nunique().count()) + ' neighborhoods.')
print('Total number of neighborhoods in Toronto is ' + str(toronto_data.shape[0]))

2268 venues found in 101 neighborhoods.
Total number of neighborhoods in Toronto is 103


There are no nearby venues for 2 neighborhoods. Let's take a look at the dataframe.

In [24]:
grouped_toronto_venues = toronto_venues.groupby('Neighborhood').count().reset_index()
grouped_toronto_venues

Neighborhood  Neighborhood Latitude  \
0                                            Agincourt                      5   
1                               Alderwood, Long Branch                      9   
2      Bathurst Manor, Wilson Heights, Downsview North                     20   
3                                      Bayview Village                      2   
4                    Bedford Park, Lawrence Manor East                     20   
5                                          Berczy Park                     67   
6                          Birch Cliff, Cliffside West                      4   
7         Brockton, Parkdale Village, Exhibition Place                     43   
8                Business reply mail Processing CentrE                    100   
9    CN Tower, King and Spadina, Railway Lands, Har...                     66   
10                                 Caledonia-Fairbanks                      8   
11               Canada Post Gateway Processing Centre                    100   
12                                           Cedarbrae                      1   
13                                  Central Bay Street                     57   
14                                            Christie                     12   
15                                Church and Wellesley                     79   
16             Clarks Corners, Tam O'Shanter, Sullivan                      7   
17     Cliffside, Cliffcrest, Scarborough Village West                      5   
18                      Commerce Court, Victoria Hotel                    100   
19                                          Davisville                     26   
20                                    Davisville North                      5   
21    Del Ray, Mount Dennis, Keelsdale and Silverthorn                      4   
22                                       Don Mills M3B                      6   
23                                       Don Mills M3C                      4   
24   Dorset Park, Wexford Heights, Scarborough Town...                      6   
25                                       Downsview M3K                      4   
26                                       Downsview M3L                      2   
27                                       Downsview M3M                      1   
28                                       Downsview M3N                     21   
29                        Dufferin, Dovercourt Village                     16   
30   Eringate, Bloordale Gardens, Old Burnhamthorpe...                     10   
31                        Fairview, Henry Farm, Oriole                     53   
32              First Canadian Place, Underground city                    100   
33                            Forest Hill North & West                      1   
34                            Garden District, Ryerson                    100   
35                                           Glencairn                      7   
36                     Golden Mile, Clairlea, Oakridge                      8   
37                   Guildwood, Morningside, West Hill                     20   
38   Harbourfront East, Union Station, Toronto Islands                      7   
39                       High Park, The Junction South                      5   
40                                   Hillcrest Village                      3   
41                                       Humber Summit                      4   
42                                    Humberlea, Emery                      2   
43                                  Humewood-Cedarvale                      6   
44                      India Bazaar, The Beaches West                     20   
45                                    Islington Avenue                      4   
46         Kennedy Park, Ionview, East Birchmount Park                      4   
47           Kensington Market, Chinatown, Grange Park                     43   
48   Kingsview Village, St. Phillips, Martin Grove ...                   

And take a look at the categories

In [25]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 268 uniques categories.


There are too few venues in some neighborhoods to perform a significant cluster analysis. I will exclude neighborhoods with less then 20 venues.

In [26]:
too_few = grouped_toronto_venues[grouped_toronto_venues['Venue'] < 20].reset_index()
print('There are ' + str(too_few.shape[0]) + ' neighborhoods with less than 20 nearby venues')

There are 68 neighborhoods with less than 20 nearby venues


In [27]:
indexes = toronto_venues[toronto_venues["Neighborhood"].isin(too_few['Neighborhood'])].index

toronto_venues_filtered = toronto_venues.drop(indexes)
toronto_venues_filtered.reset_index(inplace=True, drop=True)
toronto_venues_filtered.shape

(1888, 7)

In [28]:
toronto_venues_filtered.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront              43.650964              -79.353041   
1  Regent Park, Harbourfront              43.650964              -79.353041   
2  Regent Park, Harbourfront              43.650964              -79.353041   
3  Regent Park, Harbourfront              43.650964              -79.353041   
4  Regent Park, Harbourfront              43.650964              -79.353041   

                                  Venue  Venue Latitude  Venue Longitude  \
0                           Souk Tabule       43.653756       -79.354390   
1  Young Centre for the Performing Arts       43.650825       -79.357593   
2                   SOMA chocolatemaker       43.650622       -79.358127   
3            Cluny Bistro & Boulangerie       43.650565       -79.357843   
4                           BATLgrounds       43.647088       -79.351306   

             Venue Category  
0  Mediterranean Restaurant  
1     Performing Arts Venue  
2            Chocolate Shop  
3         French Restaurant  
4        Athletics & Sports

In [29]:
# do the same for toronto_data

toronto_data_filtered = toronto_data.copy()

indexes = toronto_data_filtered[toronto_data_filtered["Neighborhood"].isin(too_few['Neighborhood'])].index
toronto_data_filtered.drop(indexes , inplace=True)
toronto_data_filtered.reset_index(inplace=True, drop=True)
toronto_data_filtered.shape

#  there area still 2 neighborhoods for which the search returned no values...

(35, 5)

Now let's check if neighborhoods with too few venues have beeen correctly deleted

In [31]:
toronto_venues_filtered.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Bathurst Manor, Wilson Heights, Downsview North                        20   
Bedford Park, Lawrence Manor East                                      20   
Berczy Park                                                            67   
Brockton, Parkdale Village, Exhibition Place                           43   
Business reply mail Processing CentrE                                 100   
CN Tower, King and Spadina, Railway Lands, Harb...                     66   
Canada Post Gateway Processing Centre                                 100   
Central Bay Street                                                     57   
Church and Wellesley                                                   79   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             26   
Downsview M3N                                                          21   
Fairview, Henry Farm, Oriole                                           53   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Guildwood, Morningside, West Hill                                      20   
India Bazaar, The Beaches West                                         20   
Kensington Market, Chinatown, Grange Park                              43   
Lawrence Manor, Lawrence Heights                                       52   
Leaside                                                                37   
Little Portugal, Trinity                                               43   
Parkdale, Roncesvalles                                                 66   
Queen's Park, Ontario Provincial Government                            41   
Regent Park, Harbourfront                                              26   
Richmond, Adelaide, King                                               98   
Runnymede, Swansea                                                     50   
St. James Town                                                         82   
St. James Town, Cabbagetown                                            47   
Stn A PO Boxes                                                        100   
The Annex, North Midtown, Yorkville                                    26   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   
Willowdale M2N                                                         51   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Bathurst Manor, Wilson Heights, Downsview North                         20   
Bedford Park, Lawrence Manor East                                       20   
Berczy Park                                                             67   
Brockton, Parkdale Village, Exhibition Place                            43   
Business reply mail Processing CentrE                                  100   
CN Tower, King and Spadina, Railway Lands, Harb...                      66   
Canada Post Gateway Processing Centre                                  100   
Central Bay Street                                                      57   
Church and Wellesley                                                    79   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              26   
Downsview M3N                                                           21   
Fairview, Henry Farm, Oriole                                            53   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Guildwood, Morningside, West H

Create a new map only with selected neighborhoods in new dataframe.
Same longitude and latitiude as before, a little more zoom.

In [32]:
# create map of Toronto using latitude and longitude values
map_toronto_filtered = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data_filtered['Latitude'], toronto_data_filtered['Longitude'], toronto_data_filtered['Borough'], toronto_data_filtered['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto_filtered)  
    
map_toronto_filtered

Data preprocessing

In [33]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues_filtered[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues_filtered['Neighborhood'] 

toronto_onehot.shape

(1888, 227)

In [34]:
# move neighborhood column to the first column
nb = toronto_onehot.columns.get_loc('Neighborhood')
fixed_columns = [toronto_onehot.columns[nb]] + list(toronto_onehot.columns[: nb]) + list(toronto_onehot.columns[nb+1 :]) 

toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood  Accessories Store  American Restaurant  \
0  Regent Park, Harbourfront                  0                    0   
1  Regent Park, Harbourfront                  0                    0   
2  Regent Park, Harbourfront                  0                    0   
3  Regent Park, Harbourfront                  0                    0   
4  Regent Park, Harbourfront                  0                    0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0            0           0                    0                 0   
1            0           0                    0                 0   
2            0           0                    0                 0   
3            0           0                    0                 0   
4            0           0                    0                 0   

   Athletics & Sports  Auto Dealership  Auto Workshop  BBQ Joint  Bagel Shop  \
0                   0                0              0          0           0   
1                   0                0              0          0           0   
2                   0                0              0          0           0   
3                   0                0              0          0           0   
4                   1                0              0          0           0   

   Bakery  Bank  Bar  Baseball Field  Basketball Stadium  Beach  Beach Bar  \
0       0     0    0               0                   0      0          0   
1       0     0    0               0                   0      0          0   
2       0     0    0               0                   0      0          0   
3       0     0    0               0                   0      0          0   
4       0     0    0               0                   0      0          0   

   Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  Bistro  Bookstore  \
0         0           0                   0          0       0          0   
1         0           0                   0          0       0          0   
2         0           0                   0          0       0          0   
3         0           0                   0          0       0          0   
4         0           0                   0          0       0          0   

   Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0         0                     0               0        0            0   
1         0                     0               0        0            0   
2         0                     0               0        0            0   
3         0                     0               0        0            0   
4         0                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Business Service  \
0                0         0             0              0                 0   
1                0         0             0              0                 0   
2                0         0             0              0                 0   
3                0         0             0              0                 0   
4                0         0             0              0                 0   

   Butcher  Café  Caribbean Restaurant  Cheese Shop  Chinese Restaurant  \
0        0     0                     0            0                   0   
1        0     0                     0            0                   0   
2        0     0                     0            0                   0   
3        0     0                     0            0                   0   
4        0     0                     0            0                   0   

   Chiropractor  Chocolate Shop  Church  Clothing Store  Cocktail Bar  \
0             0               0       0               0             0   
1             0               0       0               0             0   
2             0               1       0               0             0   
3             0               0       0               0             0   
4            

And let's examine the new dataframe size.

In [35]:
toronto_onehot.shape

(1888, 227)

Calculate the mean of the occurrence of each category grouped by neighborhood

In [36]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Accessories Store  \
0     Bathurst Manor, Wilson Heights, Downsview North           0.000000   
1                   Bedford Park, Lawrence Manor East           0.000000   
2                                         Berczy Park           0.000000   
3        Brockton, Parkdale Village, Exhibition Place           0.023256   
4               Business reply mail Processing CentrE           0.000000   
5   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
6               Canada Post Gateway Processing Centre           0.000000   
7                                  Central Bay Street           0.000000   
8                                Church and Wellesley           0.000000   
9                      Commerce Court, Victoria Hotel           0.000000   
10                                         Davisville           0.000000   
11                                      Downsview M3N           0.000000   
12                       Fairview, Henry Farm, Oriole           0.000000   
13             First Canadian Place, Underground city           0.000000   
14                           Garden District, Ryerson           0.000000   
15                  Guildwood, Morningside, West Hill           0.000000   
16                     India Bazaar, The Beaches West           0.000000   
17          Kensington Market, Chinatown, Grange Park           0.000000   
18                   Lawrence Manor, Lawrence Heights           0.000000   
19                                            Leaside           0.000000   
20                           Little Portugal, Trinity           0.000000   
21                             Parkdale, Roncesvalles           0.000000   
22        Queen's Park, Ontario Provincial Government           0.000000   
23                          Regent Park, Harbourfront           0.000000   
24                           Richmond, Adelaide, King           0.000000   
25                                 Runnymede, Swansea           0.000000   
26                                     St. James Town           0.000000   
27                        St. James Town, Cabbagetown           0.000000   
28                                     Stn A PO Boxes           0.000000   
29                The Annex, North Midtown, Yorkville           0.000000   
30           Toronto Dominion Centre, Design Exchange           0.000000   
31                     University of Toronto, Harbord           0.000000   
32                                     Willowdale M2N           0.000000   

    American Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0              0.000000     0.000000    0.000000             0.000000   
1              0.000000     0.000000    0.000000             0.000000   
2              0.000000     0.014925    0.000000             0.000000   
3              0.000000     0.023256    0.000000             0.023256   
4              0.030000     0.010000    0.000000             0.010000   
5              0.000000     0.000000    0.000000             0.000000   
6              0.030000     0.010000    0.000000             0.010000   
7              0.000000     0.000000    0.017544             0.000000   
8              0.012658     0.000000    0.000000             0.012658   
9              0.040000     0.010000    0.000000             0.000000   
10             0.000000     0.000000    0.000000             0.000000   
11             0.000000     0.000000    0.000000             0.000000   
12             0.000000     0.000000    0.000000             0.000000   
13             0.040000     0.010000    0.000000             0.000000   
14             0.010000     0.010000    0.000000             0.000000   
15             0.000000     0.000000    0.000000             0.000000   
16             0.000000     0.000000    0.000000             0.000000   
17             0.000000     0.023256    0.000000             0.023256   
18             0.038462     0.000000    0.000000             0.000000

In [37]:
toronto_grouped.shape

(33, 227)

Function for returning most common values

In [38]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [39]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
1                Bedford Park, Lawrence Manor East           Coffee Shop   
2                                      Berczy Park           Coffee Shop   
3     Brockton, Parkdale Village, Exhibition Place           Coffee Shop   
4            Business reply mail Processing CentrE           Coffee Shop   

  2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
0                  Bank                Pharmacy           Gas Station   
1        Sandwich Place      Italian Restaurant              Pharmacy   
2    Seafood Restaurant            Cocktail Bar    Italian Restaurant   
3                  Café  Thrift / Vintage Store             Gift Shop   
4                 Hotel                    Café   American Restaurant   

  5th Most Common Venue    6th Most Common Venue 7th Most Common Venue  \
0             Pet Store              Pizza Place         Deli / Bodega   
1       Thai Restaurant  Comfort Food Restaurant           Pizza Place   
2        Farmers Market                   Bakery                  Café   
3                 Diner           Sandwich Place  Caribbean Restaurant   
4      Asian Restaurant               Restaurant   Japanese Restaurant   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0                 Diner            Restaurant  Middle Eastern Restaurant  
1                   Pub            Restaurant                       Café  
2        Breakfast Spot            Restaurant                      Hotel  
3          Chiropractor            Restaurant               Cocktail Bar  
4       Thai Restaurant            Steakhouse                        Gym

### 3.2 Cluster Analysis

Run *k*-means to cluster the neighborhood into 5 clusters.

In [40]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([1, 2, 0, 2, 0, 2, 0, 0, 0, 0, 2, 1, 3, 0, 0, 1, 2, 2, 3, 2, 2, 2,
       4, 2, 0, 2, 0, 2, 0, 2, 0, 2, 2], dtype=int32)

In [41]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data_filtered.drop('PostalCode',1)

# merge toronto_grouped with toronto_data_filtered to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged

Borough                                       Neighborhood  \
0   Downtown Toronto                          Regent Park, Harbourfront   
1         North York                   Lawrence Manor, Lawrence Heights   
2   Downtown Toronto        Queen's Park, Ontario Provincial Government   
3   Downtown Toronto                           Garden District, Ryerson   
4   Downtown Toronto                                     St. James Town   
5        Scarborough                  Guildwood, Morningside, West Hill   
6   Downtown Toronto                                        Berczy Park   
7          East York                                            Leaside   
8   Downtown Toronto                                 Central Bay Street   
9         North York    Bathurst Manor, Wilson Heights, Downsview North   
10  Downtown Toronto                           Richmond, Adelaide, King   
11        North York                       Fairview, Henry Farm, Oriole   
12         East York                                       East Toronto   
13      West Toronto                           Little Portugal, Trinity   
14  Downtown Toronto           Toronto Dominion Centre, Design Exchange   
15      West Toronto       Brockton, Parkdale Village, Exhibition Place   
16      East Toronto                     India Bazaar, The Beaches West   
17  Downtown Toronto                     Commerce Court, Victoria Hotel   
18        North York                  Bedford Park, Lawrence Manor East   
19        North York                                     Willowdale M2N   
20        North York                                      Downsview M3N   
21   Central Toronto                The Annex, North Midtown, Yorkville   
22      West Toronto                             Parkdale, Roncesvalles   
23       Mississauga              Canada Post Gateway Processing Centre   
24   Central Toronto                                         Davisville   
25  Downtown Toronto                     University of Toronto, Harbord   
26      West Toronto                                 Runnymede, Swansea   
27  Downtown Toronto          Kensington Market, Chinatown, Grange Park   
28  Downtown Toronto  CN Tower, King and Spadina, Railway Lands, Har...   
29  Downtown Toronto                                     Stn A PO Boxes   
30       Scarborough                                        Upper Rouge   
31  Downtown Toronto                        St. James Town, Cabbagetown   
32  Downtown Toronto             First Canadian Place, Underground city   
33  Downtown Toronto                               Church and Wellesley   
34      East Toronto              Business reply mail Processing CentrE   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0   43.650964 -79.353041             2.0                   Pub   
1   43.723265 -79.451211             3.0        Clothing Store   
2   43.661790 -79.389390             4.0           Coffee Shop   
3   43.657491 -79.377529             0.0           Coffee Shop   
4   43.651734 -79.375554             0.0           Coffee Shop   
5   43.765806 -79.185284             1.0  Fast Food Restaurant   
6   43.645196 -79.373855             0.0           Coffee Shop   
7   43.709414 -79.363100             2.0           Coffee Shop   
8   43.656072 -79.385653             0.0           Coffee Shop   
9   43.757394 -79.442394             1.0           Coffee Shop   
10  43.650542 -79.384116             0.0           Coffee Shop   
11  43.780306 -79.348687             3.0        Clothing Store   
12  43.687046 -79.333890             NaN                   NaN   
13  43.648636 -79.417752             2.0            Restaurant   
14  43.646923 -79.381626             0.0           Coffee Shop   
15  43.639922 -79.431240             2.0           Coffee Shop   
16  43.668291 -79.315578             2.0        Sandwich Place   
17  43.648283 -79.378750             0.0           Coffee Shop   
18  43.735447 -79.417944             2.0           Coffee 

In [42]:
# the 2 neighborhoods with no venues are still here...

toronto_merged.dropna(inplace=True)

# Cluster Labels colums was implicitly converted to float when we made the join
toronto_merged['Cluster Labels'] = toronto_merged['Cluster Labels'].astype(int)

In [43]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Cluster 1

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
3                   Garden District, Ryerson           Coffee Shop   
4                             St. James Town           Coffee Shop   
6                                Berczy Park           Coffee Shop   
8                         Central Bay Street           Coffee Shop   
10                  Richmond, Adelaide, King           Coffee Shop   
14  Toronto Dominion Centre, Design Exchange           Coffee Shop   
17            Commerce Court, Victoria Hotel           Coffee Shop   
23     Canada Post Gateway Processing Centre           Coffee Shop   
29                            Stn A PO Boxes           Coffee Shop   
32    First Canadian Place, Underground city           Coffee Shop   
33                      Church and Wellesley           Coffee Shop   
34     Business reply mail Processing CentrE           Coffee Shop   

   2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
3         Clothing Store  Middle Eastern Restaurant            Restaurant   
4                   Café         Italian Restaurant        Cosmetics Shop   
6     Seafood Restaurant               Cocktail Bar    Italian Restaurant   
8    Japanese Restaurant  Middle Eastern Restaurant                  Café   
10                  Café             Clothing Store            Restaurant   
14                 Hotel                       Café            Restaurant   
17                  Café                      Hotel            Restaurant   
23                 Hotel                       Café   American Restaurant   
29                 Hotel                       Café   American Restaurant   
32                  Café                      Hotel            Restaurant   
33   Japanese Restaurant                 Restaurant                   Pub   
34                 Hotel                       Café   American Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
3         Sandwich Place        Cosmetics Shop                 Hotel   
4    American Restaurant             Gastropub                 Hotel   
6         Farmers Market                Bakery                  Café   
8        Bubble Tea Shop        Clothing Store                 Plaza   
10                   Gym   American Restaurant         Deli / Bodega   
14   Japanese Restaurant   American Restaurant           Salad Place   
17   American Restaurant   Japanese Restaurant                   Gym   
23      Asian Restaurant            Restaurant   Japanese Restaurant   
29      Asian Restaurant            Restaurant   Japanese Restaurant   
32   American Restaurant         Deli / Bodega                   Gym   
33      Sushi Restaurant          Dance Studio           Men's Store   
34      Asian Restaurant            Restaurant   Japanese Restaurant   

   8th Most Common Venue     9th Most Common Venue 10th Most Common Venue  
3                    Bar                      Café                  Diner  
4     Seafood Restaurant              Cocktail Bar       Department Store  
6         Breakfast Spot                Restaurant                  Hotel  
8         Breakfast Spot            Sandwich Place             Restaurant  
10                 Hotel               Salad Place       Sushi Restaurant  
14    Seafood Restaurant                  Tea Room     Italian Restaurant  
17         Deli / Bodega                  Beer Bar     Seafood Restaurant  
23       Thai Restaurant                Steakhouse                    Gym  
29       Thai Restaurant                Steakhouse                    Gym  
32   Japanese Restaurant          Asian Restaurant                    Bar  
33                  Café  Mediterranean Restaurant           Burger Joint  
34       Thai Restaurant                Steakhouse                    Gym

#### Cluster 2

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
5                 Guildwood, Morningside, West Hill  Fast Food Restaurant   
9   Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
20                                    Downsview M3N         Shopping Mall   

   2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
5            Pizza Place  Thrift / Vintage Store   Fried Chicken Joint   
9                   Bank                Pharmacy           Gas Station   
20         Grocery Store             Pizza Place        Discount Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5             Restaurant            Beer Store           Coffee Shop   
9              Pet Store           Pizza Place         Deli / Bodega   
20  Fast Food Restaurant          Liquor Store   Fried Chicken Joint   

   8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
5             Laundromat            Sports Bar             Breakfast Spot  
9                  Diner            Restaurant  Middle Eastern Restaurant  
20           Coffee Shop           Gas Station          Mobile Phone Shop

#### Cluster 3

In [46]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
0                           Regent Park, Harbourfront                   Pub   
7                                             Leaside           Coffee Shop   
13                           Little Portugal, Trinity            Restaurant   
15       Brockton, Parkdale Village, Exhibition Place           Coffee Shop   
16                     India Bazaar, The Beaches West        Sandwich Place   
18                  Bedford Park, Lawrence Manor East           Coffee Shop   
19                                     Willowdale M2N      Ramen Restaurant   
21                The Annex, North Midtown, Yorkville        Sandwich Place   
22                             Parkdale, Roncesvalles           Coffee Shop   
24                                         Davisville          Dessert Shop   
25                     University of Toronto, Harbord                  Café   
26                                 Runnymede, Swansea                  Café   
27          Kensington Market, Chinatown, Grange Park                  Café   
28  CN Tower, King and Spadina, Railway Lands, Har...           Coffee Shop   
31                        St. James Town, Cabbagetown                  Café   

   2nd Most Common Venue        3rd Most Common Venue  4th Most Common Venue  \
0            Music Venue                  Coffee Shop                   Café   
7    Sporting Goods Shop                   Sports Bar           Burger Joint   
13          Cocktail Bar                          Bar               Wine Bar   
15                  Café       Thrift / Vintage Store              Gift Shop   
16  Fast Food Restaurant                         Park       Sushi Restaurant   
18        Sandwich Place           Italian Restaurant               Pharmacy   
19           Coffee Shop            Korean Restaurant                   Park   
21                  Café                     Pharmacy     Italian Restaurant   
22         Grocery Store  Eastern European Restaurant       Sushi Restaurant   
24                  Café               Sandwich Place            Coffee Shop   
25    Italian Restaurant                   Restaurant                   Park   
26           Coffee Shop             Sushi Restaurant           Dessert Shop   
27    Mexican Restaurant                  Coffee Shop  Vietnamese Restaurant   
28            Restaurant                         Café      French Restaurant   
31            Restaurant                  Coffee Shop                   Park   

    5th Most Common Venue          6th Most Common Venue  \
0      Athletics & Sports                           Bank   
7                    Bank         Furniture / Home Store   
13  Vietnamese Restaurant               Asian Restaurant   
15                  Diner                 Sandwich Place   
16                    Pub                    Coffee Shop   
18        Thai Restaurant        Comfort Food Restaurant   
19         Sandwich Place                    Pizza Place   
21         Cosmetics Shop                    Pizza Place   
22             Restaurant                Thai Restaurant   
24            Pizza Place             Italian Restaurant   
25              Bookstore            Japanese Restaurant   
26            Pizza Place             Falafel Restaurant   
27          Grocery Store  Vegetarian / Vegan Restaurant   
28                   Park                      Speakeasy   
31     Italian Restaurant                            Pub   

   7th Most Common Venue 8th Most Common Venue       9th Most Common Venue  \
0                 Bakery          Tech Startup             Thai Restaurant   
7          Shopping Mall               Brewery              Breakfast Spot   
13           Coffee Shop           Yoga Studio                Cupcake Shop   
15  Caribbean Restaurant          Chiropractor                  Restaurant   
16    Italian Restaurant         Movie Theater                  Steakhouse   
18           Pizza Place                   Pub                  Re

#### Cluster 4

In [47]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
1   Lawrence Manor, Lawrence Heights        Clothing Store   
11      Fairview, Henry Farm, Oriole        Clothing Store   

   2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
1       Toy / Game Store   American Restaurant  Furniture / Home Store   
11            Restaurant           Coffee Shop    Fast Food Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1            Men's Store             Bookstore            Restaurant   
11            Food Court         Women's Store      Toy / Game Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1         Cosmetics Shop         Women's Store            Coffee Shop  
11                  Bank              Boutique              Juice Bar

#### Cluster 5

In [48]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
2  Queen's Park, Ontario Provincial Government           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2                  Café      Sushi Restaurant           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2                  Park                   Spa          Burger Joint   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
2         Burrito Place        Sandwich Place     College Auditorium